data/strings/sound.txt

```diff
@@ -36,9 +35,17 @@
     tr = Sağdan devam edin.
     uk = Тримайтеся правіше.
     vi = Đi về phía phải.
-    zh-Hans = 向右走。
+    zh-Hans = 向右行驶
     zh-Hant = 向右走。
 
+  [make_a_slight_right_turn_street]
+    comment = Optionally used inside dist_direction_onto_street in position 3 when the normal make_a_slight_right_turn is insufficient.
+    nl = de flauwe bocht naar rechts
+
+  [make_a_slight_right_turn_street_verb]
+    comment = Optionally used inside dist_direction_onto_street in position 5 when make_a_slight_right_turn_street needs to be split into two parts.
```

Comment:
@1ec5 thanks for the review Minh! We are using format strings in the sense of `Turn $1 onto $2` etc which is about as flexible as we can get without me majorly rearchitecting the entire app's i18n as my first (long delayed) PR.

We're moving towards Weblate as fast as we can but the core issue is that the developer workflow is via these text files which are actually not a supported format. They look like INIs but most platforms don't support all strings (all languages) in one file: they expect one file per language. So the most direct fix is to just abandon the txt file format entirely and change the per-platform string files directly... but that's a big change for developers and core contributors used to doing it the other way for so long.

Reply:
> We are using format strings in the sense of `Turn $1 onto $2` etc which is about as flexible as we can get without me majorly rearchitecting the entire app's i18n as my first (long delayed) PR.

“Turn $1 onto $2” would be fine, but the current approach seems to be literally “$1 $2 $3 $4”, where $1 is “Turn” and $3 is “onto”. That’s no different than concatenating Lego strings together. Placeholders are for dynamic content, e.g., street names, not static strings. The challenge OSRMTI faced was that we had to choose between treating words like “right” as dynamic content (harder for some languages to integrate into the overall sentence without custom code) or inlining them as static content (resulting in a proliferation of format strings, because these words can’t be reused as easily). The current approach really leans into the former.

If you’re open to addressing this issue as part of this PR, I think it would be a matter of inlining the existing translations into the format string as a one-time operation. Then the issue of null slots goes away, because the translator can simply omit a word in the format string, confident that another language’s translation won’t take its place.

Reply:
@1ec5 I hear you, but I think you're underestimating the complexity of the format string -- which does change, as some languages have different subject-verb-adjective-object ordering. We could certainly move some logic out of the locale strings and into the program, but honestly the most stable fix would be to just have some locale strings/files behave differently than others (maybe a marker or argument that disables the english fallback logic etc.)

Merging some of these strings into the formatter to remove the formatter would likely increase duplication not decrease: previous versions of this PR had something like `$1 $2 onto $3` until we realized that "onto" actually needs to itself be dynamic based on a few different factors.

It's true that we can't very easily reuse any part of "take the fifth exit" across all languages due to counting issues, what constitutes an "exit," how to conjugate "take", etc, but that's not the biggest problem the app faces so I'm not super worried. It is a future goal to support arbitrary distances and countings, like "in `13` meters, turn right", which will put all of this to the test especially with that fun fun fun pluralization, and that's being discussed and maybe we do move to something like Fluent later -- cc @pastk 

For now, this PR is about including the street name in the turn instructions, and I fear if we stray past the minimal changeset to accomplish that goal then it'll be 2026 before this feature gets released.

Reply:
> I hear you, but I think you're underestimating the complexity of the format string -- which does change, as some languages have different subject-verb-adjective-object ordering.

I don’t think I am. “$1 $2 onto $3” can be translated as “Onto $3 by $2ing $1” if necessary. What you’d lose is the ability to tell that “onto” is the preposition that needs to be inclined differently depending on $3’s gender, so you’d have to have a different string per gender or some form of markup within the string itself.

> Merging some of these strings into the formatter to remove the formatter would likely increase duplication not decrease: previous versions of this PR had something like `$1 $2 onto $3` until we realized that "onto" actually needs to itself be dynamic based on a few different factors.
> 
> It's true that we can't very easily reuse any part of "take the fifth exit" across all languages due to counting issues, what constitutes an "exit," how to conjugate "take", etc, but that's not the biggest problem the app faces so I'm not super worried.

Right, my point is that redundant, unreusable strings aren’t necessarily such a problem that you need to avoid them so aggressively. It’s OK to make translators translate more strings if that gives them more control over the final output. OSRMTI struck a balance that translators found somewhat daunting at first, but it isn’t nearly as burdensome as in Valhalla, which does minimal substitution or concatenation. Sometimes OSRMTI contributors would complain about having to translate 300 strings, but once they got started on the first 20%, Transifex’s translation memory kicked in, speeding up the remaining 80%.

Fluent (and a work-in-progress industry standard, [MessageFormat2](https://github.com/unicode-org/message-format-wg/)) would somewhat address this tradeoff, so that translators would work with something streamlined and the redundant strings would only occur in the built product. Translatewiki.net supports an even more powerful templating language as part of its message format, but that’s less useful for mobile platforms where there’s no tooling to interpret that language.

It isn’t for me to insist on reworking the translation units as part of this PR. Going forward, there will be plenty of opportunities to experience the limitations of the current approach and evaluate alternatives (not that the alternatives are unalloyed goods either).

Reply:
The instant we get any more complex than this I fully agree; as you can see Hungarian was a whole thing by itself and I'm not sure Fluent helps much with vowel harmony especially for unknown variables (street names).

But like I said there is a desire to support dynamic numbers and that's a prime Fluent use case; we're already in the realm of silliness with 11 strings for 1-11 roundabout exits. I see the value, but I think it's a project for a future day